plugins/modules/foreman_hostgroup.py

```diff
@@ -220,22 +220,39 @@
 RETURN = ''' # '''
 
 from ansible.module_utils.foreman_helper import (
-    build_fqn,
     HostMixin,
     ForemanTaxonomicEntityAnsibleModule,
     OrganizationMixin,
     parameter_entity_spec,
-    split_fqn,
 )
 
 
-class ForemanHostgroupAnsibleModule(OrganizationMixin, HostMixin, ForemanTaxonomicEntityAnsibleModule):
+class ForemanHostgroupModule(OrganizationMixin, HostMixin, ForemanTaxonomicEntityAnsibleModule):
     pass
 
 
+def ensure_puppetclasses(module, entity):
+    entity_dict = module.clean_params()
+    expected_puppetclasses = entity_dict.pop('puppetclasses', None)
+    if expected_puppetclasses:
+        expected_puppetclasses = module.find_puppetclasses(expected_puppetclasses, environment=entity['environment_id'], thin=True)
+    current_puppetclasses = entity.pop('puppetclass_ids', [])
+    if expected_puppetclasses:
+        for puppetclass in expected_puppetclasses:
+            if puppetclass['id'] in current_puppetclasses:
+                current_puppetclasses.remove(puppetclass['id'])
+            else:
+                payload = {'hostgroup_id': entity['id'], 'puppetclass_id': puppetclass['id']}
+                module.ensure_entity('hostgroup_classes', {}, None, params=payload, state='present', entity_spec={})
+        if len(current_puppetclasses) > 0:
+            for leftover_puppetclass in current_puppetclasses:
+                module.ensure_entity('hostgroup_classes', {}, {'id': leftover_puppetclass}, {'hostgroup_id': entity['id']}, state='absent', entity_spec={})
+
+
 def main():
-    module = ForemanHostgroupAnsibleModule(
+    module = ForemanHostgroupModule(
         entity_spec=dict(
+            organization=dict(type='entity', flat_name='organization_id', ensure=False),
```

Comment:
this `ensure=False` kinda feels like "we put this into the entity_spec, but it's not really part of the entity" (and that matches the old code, we had org and puppetclasses in the regular argument spec). I wonder if that's really worth the added complexity.

Reply:
Only organization was in the argument spec, puppetclasses was already there and popped before ensure call in the module. For the organization part, I think it was in arg_spec instead of entity_spec because hostgroup module have to handle 'bare' foreman workflow and katello workflow (which needs an org). It's the only module that do that. I understand your point but put it in entity like this will keep on 'declarative way' instead of writing code on module side. I can add a comment before this line to explain or fallback to the 'old way', or keep it 'as is' as you want.

Reply:
Yeah, it's there for Katello-related params (CV etc), which is scoped to a single org.

I think my main source of dislike comes from the fact that `foreman_hostgroup` is the only module that uses `ensure=False` (for puppetclasses and org), which makes me thing this special handling should happen in the module, not in the helper (so "old way" if you want it). But then again the old way was not 100% consistent too (as you correctly pointed out).

Slight preference to put both into `arg_spec` and let the module handle it.

@mdellweg?

Reply:
Do we need `ensure=False` or can we take it out of the `entity_dict` after all the lookup and pass that to `run`?

Reply:
Reverted that and manage it before calling `run()`